calyx/src/passes/group_to_seq.rs

```diff
@@ -0,0 +1,380 @@
+    }
+}
+
+impl Visitor for GroupToSeq {
+    fn start(
+        &mut self,
+        comp: &mut ir::Component,
+        sigs: &ir::LibrarySignatures,
+        _comps: &[ir::Component],
+    ) -> VisResult {
+        let groups: Vec<ir::RRC<ir::Group>> = comp.groups.drain().collect();
+        let mut builder = ir::Builder::new(comp, sigs);
+        for g in groups.iter() {
+            let mut group = g.borrow_mut();
+            match SplitAnalysis::get_split(
+                group.assignments.drain(..).collect::<Vec<ir::Assignment>>(),
+                group.clone_name(),
+                &mut builder,
+            ) {
+                Ok((group1, group2)) => {
+                    let seq = ir::Control::seq(vec![
+                        ir::Control::enable(group1),
+                        ir::Control::enable(group2),
+                    ]);
+                    self.group_seq_map.insert(group.clone_name(), seq);
+                }
+                // If we don't do the transformation, just add back the assignemnts
+                // we just drained back into group
+                Err(mut assigns) => group.assignments.append(&mut assigns),
+            }
+        }
+
+        // Add back the groups we drained at the beginning of this method, but
+        // filter out the empty groups that were split into smaller groups
+        comp.groups.append(groups.into_iter().filter(
+            |group: &ir::RRC<ir::Group>| !group.borrow().assignments.is_empty(),
+        ));
+        Ok(Action::Continue)
+    }
+
+    fn enable(
+        &mut self,
+        s: &mut ir::Enable,
+        _comp: &mut ir::Component,
+        _sigs: &ir::LibrarySignatures,
+        _comps: &[ir::Component],
+    ) -> VisResult {
+        let group_name = s.group.borrow().clone_name();
+        match self.group_seq_map.get(&group_name) {
+            None => Ok(Action::Continue),
+            Some(seq) => Ok(Action::Change(Box::new(ir::Control::clone(seq)))),
+        }
+    }
+}
+
+// For all port reads from name in assignment, returns whether all ports are either stable
+// or done.
+fn if_name_stable_or_done(assign: &ir::Assignment, name: &ir::Id) -> bool {
+    let reads = ReadWriteSet::port_reads(assign);
+    reads
+        .filter(|port_ref| port_ref.borrow().get_parent_name() == name)
+        .all(|port_ref| {
+            let atts = &port_ref.borrow().attributes;
+            atts.has("stable") || atts.has("done")
+        })
+}
+
+// Returns true if the cell is a component or a non-combinational primitive
+fn comp_or_non_comb(cell: &ir::RRC<ir::Cell>) -> bool {
+    match &cell.borrow().prototype {
+        ir::CellType::Primitive { is_comb, .. } => !*is_comb,
+        ir::CellType::Component { .. } => true,
+        _ => false,
+    }
+}
+
+//If asmt is a write to a cell named name returns Some(name).
+//If asmt is a write to a group port, returns None.
+fn writes_to_cell(asmt: &ir::Assignment) -> Option<ir::Id> {
```

Comment:
Use ReadWriteSet method instead: https://docs.calyxir.org/source/calyx/analysis/struct.ReadWriteSet.html

Reply:
iter::once is better than using a `vec![e].iter()`: https://doc.rust-lang.org/std/iter/fn.once.html